From inaky@linux.intel.com Fri Aug 25 19:38:00 2006
Message-Id: <20060826023758.659459000@sodium.jf.intel.com>
Date: Fri, 25 Aug 2006 19:35:28 -0700
From: Inaky Perez-Gonzalez <inaky.perez-gonzalez@intel.com>
To: greg@kroah.com
Cc: linux-usb-devel@lists.sourceforge.net, David Brownell <david-b@pacbell.net>, Alan Stern <stern@rowland.harvard.edu>, inaky@linux.intel.com, Inaky Perez-Gonzalez <inaky.perez-gonzalez@intel.com>
Subject: usb: deal with broken config descriptors
Content-Disposition: inline; filename=usb-deal-with-broken-config-descriptors.patch

Change usb_get_configuration() so that it is more tolerant to devices
with bad configuration descriptors (it'll make it ignore
configurations that fail to load).

Signed-off-by: Inaky Perez-Gonzalez <inaky.perez-gonzalez@intel.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>


---
 drivers/usb/core/config.c |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

--- gregkh-2.6.orig/drivers/usb/core/config.c
+++ gregkh-2.6/drivers/usb/core/config.c
@@ -475,7 +475,9 @@ int usb_get_configuration(struct usb_dev
 		if (result < 0) {
 			dev_err(ddev, "unable to read config index %d "
 			    "descriptor/%s\n", cfgno, "start");
-			goto err;
+			dev_err(ddev, "chopping to %d config(s)\n", cfgno);
+			dev->descriptor.bNumConfigurations = cfgno;
+			break;
 		} else if (result < 4) {
 			dev_err(ddev, "config index %d descriptor too short "
 			    "(expected %i, got %i)\n", cfgno,
